runtime/src/bank/builtins/core_bpf_migration/mod.rs

```diff
@@ -1100,4 +1122,508 @@ pub(crate) mod tests {
             },
         );
     }
+
+    // CPI mockup to test CPI to newly migrated programs.
+    mod cpi_mockup {
+        use {
+            solana_instruction::Instruction, solana_program_runtime::declare_process_instruction,
+        };
+
+        declare_process_instruction!(Entrypoint, 0, |invoke_context| {
+            let transaction_context = &invoke_context.transaction_context;
+            let instruction_context = transaction_context.get_current_instruction_context()?;
+
+            let target_program_id = instruction_context.get_key_of_instruction_account(0)?;
+
+            let instruction = Instruction::new_with_bytes(*target_program_id, &[], Vec::new());
+
+            invoke_context.native_invoke(instruction, &[])
+        });
+    }
+
+    enum TestPrototype<'a> {
+        Builtin(&'a BuiltinPrototype),
+        #[allow(unused)]
+        // We aren't migrating any stateless builtins right now. Uncomment if needed.
+        Stateless(&'a StatelessBuiltinPrototype),
+    }
+    impl<'a> TestPrototype<'a> {
+        fn deconstruct(&'a self) -> (&'a Pubkey, &'a CoreBpfMigrationConfig) {
+            match self {
+                Self::Builtin(prototype) => (
+                    &prototype.program_id,
+                    prototype.core_bpf_migration_config.as_ref().unwrap(),
+                ),
+                Self::Stateless(prototype) => (
+                    &prototype.program_id,
+                    prototype.core_bpf_migration_config.as_ref().unwrap(),
+                ),
+            }
+        }
+    }
+
+    // This test can't be used to the `compute_budget` program, unless a valid
+    // `compute_budget` program is provided as the replacement (source).
+    // See program_runtime::compute_budget_processor::process_compute_budget_instructions`.`
+    // It also can't test the `bpf_loader_upgradeable` program, as it's used in
+    // the SVM's loader to invoke programs.
+    // See `solana_svm::account_loader::load_transaction_accounts`.
+    #[test_case(TestPrototype::Builtin(&BUILTINS[0]); "system")]
+    #[test_case(TestPrototype::Builtin(&BUILTINS[1]); "vote")]
+    #[test_case(TestPrototype::Builtin(&BUILTINS[2]); "stake")]
+    #[test_case(TestPrototype::Builtin(&BUILTINS[3]); "bpf_loader_deprecated")]
+    #[test_case(TestPrototype::Builtin(&BUILTINS[4]); "bpf_loader")]
+    fn test_core_bpf_migration(prototype: TestPrototype) {
```

Comment:
We have some tests above that are similar in name. Can we bikeshed this a bit?

```suggestion
    fn test_migrate_builtin_e2e(prototype: TestPrototype) {
```